aten/src/ATen/core/type.cpp

```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+// flattening
+void flattenUnion(TypePtr& type, std::vector<TypePtr>& types) {
+  if (type->kind() == UnionType::Kind) {
+    for (auto inner_type : type->expect<UnionType>()->types()) {
+      flattenUnion(inner_type, types);
+    }
+  } else if (type->kind() == OptionalType::Kind) {
+    auto inner_type = type->expect<OptionalType>()->getElementType();
+    if (inner_type->kind() == UnionType::Kind
+        || inner_type->kind() == OptionalType::Kind) {
+      flattenUnion(inner_type, types);
+    } else {
+      types.emplace_back(inner_type);
+    }
+    types.emplace_back(NoneType::get());
+  } else {
+    types.emplace_back(type);
+  }
+}
+
+// Custom comparison for Union type auxiliary functions. We compare on
+// Kind (fast), then compare by the serialized representation
+// (slower, but guaranteed to break ties). We need to have the
+// TypeEqual struct for use in a `std::unordered_map` later.
+struct TypeEqual {
+  bool operator()(const TypePtr& a, const TypePtr& b) const {
+    return a->kind() == b->kind() && a->str() == b->str();;
+  }
+};
+
+// Helper function for `standardizeUnion`
+void filterDuplicateSubtypes(std::vector<TypePtr>& types) {
+  // If we call `unifyTypes(T, None)`, then `unifyTypes` returns
+  // `Optional[T]`. The `Optional` constructor calls the `Union`
+  // constructor, which calls `filterDuplicateSubtypes`...infinite
+  // recursion (until the stack explodes). So, let's just wrap
+  // `unifyTypes` to ensure that this condition doesn't happen. We
+  // only need to worry about this in the constructor
+  auto wrapper = [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+    // If we would have unified the types to an Optional
+    if ((t1 == NoneType::get()) ^ (t2 == NoneType::get())) {
+      return c10::nullopt;
+    }
+    return unifyTypes(t1, t2);
+  };
+
+  // Coalesce types and delete all duplicates. Moving from right to left
+  // through the vector, we try to unify the current element (`i`) with
+  // each element (`j`) before the "new" end of the vector (`end`).
+  // If we're able to unify the types at `types[i]` and `types[j]`, we
+  // decrement `end`, swap `types[j]` with the unified type, and
+  // break. Otherwise, we keep `end` where it is to signify that the
+  // new end of the vector hasn't shifted
+  size_t end_idx = types.size()-1;
+  for (size_t i = types.size()-1; i > 0; --i) {
+    for (size_t j = std::min(i-1, end_idx); ; --j) {
+      c10::optional<TypePtr> unified = wrapper(types[i], types[j]);
+      if (unified) {
+        types[j] = *unified;
+        --end_idx;
+        break;
+      }
+      // Break condition here so we don't get `j = 0; j = j-1` and end
+      // up with MAX_INT
+      if (j == 0) {
+        break;
+      }
+    }
+  }
+  // Cut off the vector's tail so that `end` is the real last element
+  types.erase(types.begin() + end_idx + 1, types.end());
+}
+
+void standardizeUnion(std::vector<TypePtr>& types) {
+  filterDuplicateSubtypes(types);
```

Comment:
I think we need to filterDuplicateSubtypes and then flatten, instead of the opposite, which we're doing now. Otherwise we'll end up with Union[Optional[int], Optional[float]] instead of Union[int, float, None]

Reply:
I'm not sure what you mean...if we have `Union[Optional[int], Optional[float]]`, `flattenUnion` will make it `Union[int, float, None]`, and then `filterDuplicateSubtypes` won't have any work to do, right? So `flattenUnion` should come first

Reply:
Is this good?

```python
    def test_union_redundant_arguments_are_skipped_optional(self):
        @torch.jit.script
        def fn(x: Union[int, Optional[float], Optional[int]]) -> str:
            return "foo"

        s = fn.graph

        FileCheck().check("x : Union[float, int, NoneType]")    \
                   .run(s)
```

Reply:
i think this is relying on python's filtering of duplicate subtypes and unification, not our own

Reply:
Thought I could get out of writing C++ tests ;(